### PR TITLE
execution/bal: serve BALs up to and beyond the WSP

### PR DESCRIPTION
**Pull request by taratorio**
## Why

EIP-7928 requires nodes to serve BALs for the weak subjectivity period (3533 epochs ≈ 113,056 blocks ≈ 15.7 days). Erigon only stores BALs in MDBX for the last `MaxReorgDepth` (96) blocks — the exec-stage prune (#19110) deletes everything older — so eth/71, `engine_getPayloadBodiesBy*V2` and `eth_getBlockAccessList` could not serve anything beyond the reorg window.

Full/archive nodes keep account/storage/code domain history well past the WSP (default prune distance 262,144 blocks), so older BALs don't need to be stored at all: they can be re-derived on demand by re-executing the block against historical state — the same approach already used by the receipts generator for devp2p `GetReceipts`.

## What

New `execution/bal` package:

- `RederiveBlockAccessList` — replays a full block (init system calls at access index −1, every transaction, finalize at `len(txns)`) recording I/O into a `VersionedIO`, mirroring the block builder's BAL recording sequence.
- `Regenerator` — caching wrapper around it: LRU (`BAL_LRU`, default 1024 entries ≈ 100MB), per-block-hash dedup mutex, `BAL_EXEC_CONCURRENCY` replay semaphore (default GOMAXPROCS/2), history reader positioned at the block's first txNum. The re-derived BAL's hash is verified against `header.BlockAccessListHash` before it is cached or served — the node only ever serves BALs it derived itself from canonical state; a mismatch (e.g. non-canonical block) degrades to "not available". Pruned history surfaces as `state.PrunedError`; pre-Amsterdam headers (no BAL commitment) return nil.

Wired into all three read paths, DB-first with regeneration fallback:

| Read path | Before (block older than 96) | After |
|---|---|---|
| eth/71 `GetBlockAccessLists` | `0x80` sentinel | regenerated BAL; `0x80` only beyond history |
| `engine_getPayloadBodiesBy{Hash,Range}V2` | `blockAccessList: null` | regenerated; `null` only beyond history |
| `eth_getBlockAccessList` | error 4444 | regenerated; 4444 only beyond history |

- devp2p: `eth.BlockAccessListGetter` interface, fallback inside `AnswerGetBlockAccessListsQuery`; `MultiClient` owns a `Regenerator` next to its receipts generator. The existing 2MiB/1024-lookup response limits bound per-request replay work.
- engine API: fallback in the execmodule payload-bodies getters (serves in-process and gRPC consumers); `beginOverlayOrRo` now returns `kv.TemporalTx` so the getters' existing tx is reused for regeneration.
- RPC: `BaseAPI.balRegenerator`, constructed when the engine is a full `rules.Engine`.

BAL storage and pruning are unchanged — this PR only adds the serving fallback.

## Testing

All new tests run with `t.Parallel()`:

- `execution/bal`: insert a blockgen Amsterdam chain (transfers, contract creation with SSTORE init code, multi-tx and empty blocks), prune all stored BAL rows (asserted 0 rows left), then require regenerated bytes byte-equal blockgen's independently computed canonical BALs and their hash equal the header commitment; pre-Amsterdam returns nil.
- `p2p/protocols/eth`: handler fallback unit tests (stored wins, regenerated served, getter empty/error/unknown → sentinel, getter not consulted when stored).
- `execution/execmodule`: payload bodies serve stored BALs, then after pruning regenerate identical bytes via both ByHash and ByRange.
- `rpc/jsonrpc`: `eth_getBlockAccessList` returns the regenerated BAL after the stored rows are pruned.

`make lint` clean, full test suites of all touched packages pass, race-clean.

### CODE REVIEW

File: execution/bal/regenerator.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/erigontech/erigon/execution/exec"
+	"github.com/erigontech/erigon/execution/protocol/rules"
+	"github.com/erigontech/erigon/execution/state"
+	"github.com/erigontech/erigon/execution/types"
+)
+
+var (
+	balsCacheLimit      = dbg.EnvInt("BAL_LRU", 1024) // ~100KiB avg encoded BAL → ~100MB RAM
+	balsExecConcurrency = dbg.EnvInt("BAL_EXEC_CONCURRENCY", max(1, runtime.GOMAXPROCS(0)/2))
+)
```
[the remark this answers is not in the document]
Comment (taratorio) [reply execution/bal/regenerator.go:43]: fixed in https://github.com/erigontech/erigon/pull/21764/commits/21632c97396ea8253ce61da838cbca685c339dae

File: execution/bal/regenerator.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+// GetBlockAccessListBytes returns the canonical RLP-encoded Block Access List
+// for the given block, regenerating it via re-execution against historical
+// state. Returns (nil, nil) for blocks without a BAL commitment in the header
+// (pre-Amsterdam) and state.PrunedError-wrapped errors when the required
+// history is no longer available.
+func (g *Regenerator) GetBlockAccessListBytes(ctx context.Context, cfg *chain.Config, tx kv.TemporalTx, blockHash common.Hash, blockNum uint64) ([]byte, error) {
+	if cached, ok := g.cache.Get(blockHash); ok {
+		return cached, nil
+	}
```
[the remark this answers is not in the document]
Comment (taratorio) [reply execution/bal/regenerator.go:85]: The returned/cached bytes are deliberately shared: every caller is internal and read-only (devp2p response encoding, PayloadBody marshaling, RPC decode). This mirrors rpc/jsonrpc/receipts.Generator, which shares cached receipts with callers and explicitly documents skipping the defensive Copy() for performance. Copying ~100KiB per cache hit would defeat the point of the cache for repeated serving.

File: execution/bal/regenerator.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	computedHash := bal.Hash()
+	if computedHash != *header.BlockAccessListHash {
+		return nil, fmt.Errorf("bal.Regenerator: regenerated BAL hash mismatch for block %d (%x): got %s want %s", blockNum, blockHash, computedHash, *header.BlockAccessListHash)
+	}
+	encoded, err := types.EncodeBlockAccessListBytes(bal)
+	if err != nil {
+		return nil, err
+	}
+	g.cache.Add(blockHash, encoded)
+	return encoded, nil
```
[the remark this answers is not in the document]
Comment (taratorio) [reply execution/bal/regenerator.go:144]: The returned/cached bytes are deliberately shared: every caller is internal and read-only (devp2p response encoding, PayloadBody marshaling, RPC decode). This mirrors rpc/jsonrpc/receipts.Generator, which shares cached receipts with callers and explicitly documents skipping the defensive Copy() for performance. Copying ~100KiB per cache hit would defeat the point of the cache for repeated serving.

File: execution/bal/regenerator.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+func (m *loaderMutex[K]) lock(key K) *sync.Mutex {
+	value, _ := m.LoadOrStore(key, &sync.Mutex{})
+	mu := value.(*sync.Mutex)
+	mu.Lock()
+	return mu
+}
+
+func (m *loaderMutex[K]) unlock(mu *sync.Mutex, key K) {
+	mu.Unlock()
+	m.Delete(key)
```
[the remark this answers is not in the document]
Comment (taratorio) [reply execution/bal/regenerator.go:174]: Same loaderMutex as rpc/jsonrpc/receipts.Generator, and the race is benign: the per-key mutex is a best-effort de-dup optimization, not a correctness mechanism. A late waiter acquiring the stale mutex re-checks the cache and returns the computed entry; in the rare delete/recreate window two goroutines can replay the same block, but both derive identical canonical bytes, the cache write is idempotent, and execSem still bounds total concurrency. An airtight keyed mutex needs refcounting — not warranted for occasional duplicated work.

File: execution/bal/regenerator.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	}
+	chainReader := exec.NewChainReader(cfg, tx, g.blockReader, g.logger)
+	bal, err := RederiveBlockAccessList(ctx, cfg, g.engine, chainReader, reader, getHeader, header, block.Transactions(), block.Uncles(), block.Withdrawals(), ibs, g.logger)
+	if err != nil {
+		return nil, err
+	}
+	computedHash := bal.Hash()
+	if computedHash != *header.BlockAccessListHash {
+		return nil, fmt.Errorf("bal.Regenerator: regenerated BAL hash mismatch for block %d (%x): got %s want %s", blockNum, blockHash, computedHash, *header.BlockAccessListHash)
+	}
```
[the remark this answers is not in the document]
Comment (taratorio) [reply execution/bal/regenerator.go:138]: Addressed in https://github.com/erigontech/erigon/pull/21764/commits/e7730194c7f5c33708b07861db2d7bfdf57780b3.

**Review by yperbasis — CHANGES_REQUESTED** [review]
Requesting changes — findings below, roughly by severity.

### 1. (Medium) `engine_getPayloadBodiesByHash` hard-fails the whole batch on a regeneration mismatch — should degrade per-entry to `null`

`regenerateBlockAccessList` only degrades `state.PrunedError`; any other error — including a re-derived-hash mismatch — propagates and fails the entire `engine_getPayloadBodiesByHashV2`/`ByRange` request:

```go
if errors.Is(err, state.PrunedError) {
    return nil, nil      // degrades
}
if err != nil {
    return nil, err      // hash mismatch → fails the WHOLE batch
}
```

This is reachable: `GetPayloadBodiesByHash` resolves `number` via `HeaderNumber`, which resolves non-canonical stored headers too. The regenerator then replays the non-canonical block's txns (`BlockWithSenders(blockHash, …)`) against the canonical state for that height (`txNumReader.Min(…, blockNum)`), so the re-derived hash won't match the non-canonical header's `BlockAccessListHash` → mismatch → the whole request fails.

This contradicts the engine API spec (per-entry `null` for unavailable blocks), this PR's description ("a mismatch … degrades to not available"), the eth/71 path (which correctly swallows to the sentinel), and the pre-PR behavior (a non-canonical hash with no stored BAL returned `blockAccessList: null` for that entry, no error). The same fragility hits any transient regen error (`Header`/`BlockWithSenders`/encode) and any unforeseen mismatch on a canonical block — one bad block takes down the whole CL request.

Fix — either treat a hash mismatch (and other non-context regen errors) like `PrunedError` (return nil + `logger.Warn`, so a systematic re-derivation bug stays observable), or pre-check canonicality in the regenerator (if the canonical hash at `blockNum` != `blockHash`, return nil before re-executing — also avoids wasting a full re-execution). Please add a test: `GetPayloadBodiesByHash([nonCanonicalHash])` should return a `null` BAL entry, not an error.

### 2. (Medium) eth/71 re-execution cost vs. the response limits

`AnswerGetBlockAccessListsQuery` reuses the bodies-handler limits (`softResponseLimit` 2 MiB, `MaxBlockAccessListsServe` 1024, `lookups >= 2048`), which were sized for cheap DB lookups. With regeneration, a single request can trigger up to ~1024–2048 full historical block re-executions — and for blocks with small/empty BALs the 2 MiB size cap never bites, so the lookup cap governs. The only throttle is a per-`Regenerator` `GOMAXPROCS/2` semaphore shared across all requests, so a peer spamming these can pin CPU. This is precedented by on-demand receipts serving, but please confirm the limits are intentional given the much higher per-lookup cost — consider a separate, smaller cap on regenerations-per-request, or a global time/concurrency budget.

### 3. (Low) Three independent `Regenerator` instances

`MultiClient`, `ExecModule`, and `BaseAPI` each construct their own `bal.NewRegenerator` → up to ~300 MB of LRU (3 × `BAL_LRU` default ≈ 100 MB) and ~1.5×GOMAXPROCS concurrent replays node-wide, since each has its own semaphore. Largely unavoidable across components, but the `BAL_LRU` "≈100MB RAM" comment understates the aggregate footprint in a combined node.

### 4. (Low) `eth_getBlockAccessList` won't regenerate on a standalone rpcdaemon

`balRegenerator` is only constructed when `engine.(rules.Engine)` succeeds. A rpcdaemon talking to a remote EL over gRPC holds an `EngineReader`, so it returns 4444 for pruned blocks rather than regenerating. Same constraint as the receipts generator — worth documenting as a known gap.

### 5. (Low / nits)

- The regen path returns the LRU `[]byte` directly (the stored path clones). Fine since all consumers treat BAL bytes as read-only, but the slice is aliased across concurrent callers — a one-line caution would help future maintainers.
- `loaderMutex` delete-on-unlock can momentarily lose mutual exclusion under a specific interleaving (classic keyed-mutex flaw) → occasional duplicate re-execution of the same hash. Harmless here (idempotent recompute + hash-check + cache) and copied verbatim from `receipts.Generator`, so it's consistent — noting only.
- On a hash mismatch, `eth_getBlockAccessList` returns a generic JSON-RPC internal error rather than the 4444 "not available" used for pruned; mapping mismatch → 4444 would be consistent.

**Comment by taratorio** [comment]
> Requesting changes — findings below, roughly by severity.
> 
> ### 1. (Medium) `engine_getPayloadBodiesByHash` hard-fails the whole batch on a regeneration mismatch — should degrade per-entry to `null`
> `regenerateBlockAccessList` only degrades `state.PrunedError`; any other error — including a re-derived-hash mismatch — propagates and fails the entire `engine_getPayloadBodiesByHashV2`/`ByRange` request:
> 
> ```go
> if errors.Is(err, state.PrunedError) {
>     return nil, nil      // degrades
> }
> if err != nil {
>     return nil, err      // hash mismatch → fails the WHOLE batch
> }
> ```
> 
> This is reachable: `GetPayloadBodiesByHash` resolves `number` via `HeaderNumber`, which resolves non-canonical stored headers too. The regenerator then replays the non-canonical block's txns (`BlockWithSenders(blockHash, …)`) against the canonical state for that height (`txNumReader.Min(…, blockNum)`), so the re-derived hash won't match the non-canonical header's `BlockAccessListHash` → mismatch → the whole request fails.
> 
> This contradicts the engine API spec (per-entry `null` for unavailable blocks), this PR's description ("a mismatch … degrades to not available"), the eth/71 path (which correctly swallows to the sentinel), and the pre-PR behavior (a non-canonical hash with no stored BAL returned `blockAccessList: null` for that entry, no error). The same fragility hits any transient regen error (`Header`/`BlockWithSenders`/encode) and any unforeseen mismatch on a canonical block — one bad block takes down the whole CL request.
> 
> Fix — either treat a hash mismatch (and other non-context regen errors) like `PrunedError` (return nil + `logger.Warn`, so a systematic re-derivation bug stays observable), or pre-check canonicality in the regenerator (if the canonical hash at `blockNum` != `blockHash`, return nil before re-executing — also avoids wasting a full re-execution). Please add a test: `GetPayloadBodiesByHash([nonCanonicalHash])` should return a `null` BAL entry, not an error.
> 
> ### 2. (Medium) eth/71 re-execution cost vs. the response limits
> `AnswerGetBlockAccessListsQuery` reuses the bodies-handler limits (`softResponseLimit` 2 MiB, `MaxBlockAccessListsServe` 1024, `lookups >= 2048`), which were sized for cheap DB lookups. With regeneration, a single request can trigger up to ~1024–2048 full historical block re-executions — and for blocks with small/empty BALs the 2 MiB size cap never bites, so the lookup cap governs. The only throttle is a per-`Regenerator` `GOMAXPROCS/2` semaphore shared across all requests, so a peer spamming these can pin CPU. This is precedented by on-demand receipts serving, but please confirm the limits are intentional given the much higher per-lookup cost — consider a separate, smaller cap on regenerations-per-request, or a global time/concurrency budget.
> 
> ### 3. (Low) Three independent `Regenerator` instances
> `MultiClient`, `ExecModule`, and `BaseAPI` each construct their own `bal.NewRegenerator` → up to ~300 MB of LRU (3 × `BAL_LRU` default ≈ 100 MB) and ~1.5×GOMAXPROCS concurrent replays node-wide, since each has its own semaphore. Largely unavoidable across components, but the `BAL_LRU` "≈100MB RAM" comment understates the aggregate footprint in a combined node.
> 
> ### 4. (Low) `eth_getBlockAccessList` won't regenerate on a standalone rpcdaemon
> `balRegenerator` is only constructed when `engine.(rules.Engine)` succeeds. A rpcdaemon talking to a remote EL over gRPC holds an `EngineReader`, so it returns 4444 for pruned blocks rather than regenerating. Same constraint as the receipts generator — worth documenting as a known gap.
> 
> ### 5. (Low / nits)
> * The regen path returns the LRU `[]byte` directly (the stored path clones). Fine since all consumers treat BAL bytes as read-only, but the slice is aliased across concurrent callers — a one-line caution would help future maintainers.
> * `loaderMutex` delete-on-unlock can momentarily lose mutual exclusion under a specific interleaving (classic keyed-mutex flaw) → occasional duplicate re-execution of the same hash. Harmless here (idempotent recompute + hash-check + cache) and copied verbatim from `receipts.Generator`, so it's consistent — noting only.
> * On a hash mismatch, `eth_getBlockAccessList` returns a generic JSON-RPC internal error rather than the 4444 "not available" used for pruned; mapping mismatch → 4444 would be consistent.

@yperbasis thanks, addressed in https://github.com/erigontech/erigon/pull/21764/commits/91f7bfa5d642a299c4495c6e40fa095af7dbcbbf and https://github.com/erigontech/erigon/pull/21764/commits/3c272a2b33e8e90a10f02f847da8f5eaafedb97b. ready for re-review

**Review by yperbasis — APPROVED** [review]
Approving — all points from my previous review are addressed: per-entry degradation plus the non-canonical test, the per-request regeneration budget, standalone-rpcdaemon regeneration via `Finalize` delegation, and mismatch → 4444. The encode-before-rollback fix in `getBlockAccessLists71` also resolves a genuine pre-existing use-after-free.

Non-blocking follow-ups:

1. The `!dbg.Exec3Parallel` skip in `TestGetBlockAccessListsResponseMatrix` is justified — serial exec hard-rejects Amsterdam when `AmsterdamTime > 0` (`exec3_serial.go`), and the response-assembly matrix keeps mode-independent coverage in `handlers_test.go` — but per our skip policy it should reference a tracking issue, and none exists for the serial-exec Amsterdam gap. Please file one and link it from the skip; the precedent skips in `engine_api_bal_test.go` could point at it too.
2. Replay cancellation is only checked between transactions, so a cancelled context still runs the in-flight tx and finalization to completion while holding a regeneration slot. Consider `context.AfterFunc(ctx, evm.Cancel)` per tx and a ctx check before `FinalizeBlockExecution`, mirroring the receipts generator.
3. Nit: `eth_getBlockAccessList` gates on `data == nil` where the other two serving paths use `len(...) == 0`; `len(data) == 0` is safer against non-nil empty slices from remote-KV reads and consistent with the siblings.

**Comment by taratorio** [comment]
> Approving — all points from my previous review are addressed: per-entry degradation plus the non-canonical test, the per-request regeneration budget, standalone-rpcdaemon regeneration via `Finalize` delegation, and mismatch → 4444. The encode-before-rollback fix in `getBlockAccessLists71` also resolves a genuine pre-existing use-after-free.
> 
> Non-blocking follow-ups:
> 
> 1. The `!dbg.Exec3Parallel` skip in `TestGetBlockAccessListsResponseMatrix` is justified — serial exec hard-rejects Amsterdam when `AmsterdamTime > 0` (`exec3_serial.go`), and the response-assembly matrix keeps mode-independent coverage in `handlers_test.go` — but per our skip policy it should reference a tracking issue, and none exists for the serial-exec Amsterdam gap. Please file one and link it from the skip; the precedent skips in `engine_api_bal_test.go` could point at it too.
> 2. Replay cancellation is only checked between transactions, so a cancelled context still runs the in-flight tx and finalization to completion while holding a regeneration slot. Consider `context.AfterFunc(ctx, evm.Cancel)` per tx and a ctx check before `FinalizeBlockExecution`, mirroring the receipts generator.
> 3. Nit: `eth_getBlockAccessList` gates on `data == nil` where the other two serving paths use `len(...) == 0`; `len(data) == 0` is safer against non-nil empty slices from remote-KV reads and consistent with the siblings.

thanks, addressed:
1. N/A - the skip for serial exec is ok (tests logic only for glamsterdam parallel exec bals path)
2. done in https://github.com/erigontech/erigon/pull/21764/commits/d10c20929c0f7e92c2d428ebe481e6679d0b83dc
3. done in https://github.com/erigontech/erigon/pull/21764/commits/d10c20929c0f7e92c2d428ebe481e6679d0b83dc
